diff --git a/lib/koptocr.c b/lib/koptocr.c
index 0caab4f..98c871b 100644
--- a/lib/koptocr.c
+++ b/lib/koptocr.c
@@ -40,8 +40,8 @@ l_int32 k2pdfopt_pixGetWordBoxesInTextlines(PIX *pixs,
 void* tess_api = NULL;
 
 void k2pdfopt_tocr_init(char *datadir, char *lang) {
-	if (tess_api != NULL && strncmp(lang, k2pdfopt_tocr_get_language(tess_api), 32)) {
-		k2pdfopt_tocr_end(tess_api);
+	if (tess_api != NULL && strncmp(lang, k2pdfopt_tocr_get_language(), 32)) {
+		k2pdfopt_tocr_end();
 	}
 	if (tess_api == NULL) {
 		int status;
@@ -263,3 +263,10 @@ l_int32 k2pdfopt_pixGetWordBoxesInTextlines(PIX *pixs,
 	numaaDestroy(&naa);
 	return 0;
 }
+
+#ifdef __GNUC__
+__attribute__((destructor)) static void k2pdfopt_tocr_destructor(void) {
+	k2pdfopt_tocr_end();
+}
+#endif
+
